engine/src/main/java/org/terasology/logic/players/FirstPersonClientSystem.java

```diff
@@ -131,50 +138,61 @@ public void onHeldItemChanged(OnChangedComponent event, EntityRef character, Cha
         if (localPlayer.getCharacterEntity().equals(character)) {
             EntityRef oldHeldItem = currentHeldItem;
             currentHeldItem = heldItemComponent.selectedItem;
-            linkHeldItemLocationForLocalPlayer(character, currentHeldItem, oldHeldItem);
-        }
-    }
-
-    @ReceiveEvent(netFilter = RegisterMode.CLIENT)
-    public void onLocationRemovedAddClientSideLocation(BeforeDeactivateComponent event, EntityRef entityRef, LocationComponent locationComponent) {
-        // when in a remote client situation,  the remove LocationComponent happens after the CharacterHeldItemComponent is changed.
-        // So this allows re-adding the client side LocationComponent after this
-        if (entityRef.equals(currentHeldItem)) {
-            relinkHeldItem = true;
+            linkHeldItemLocationForLocalPlayer(currentHeldItem, oldHeldItem);
         }
     }
 
-    void linkHeldItemLocationForLocalPlayer(EntityRef character, EntityRef newItem, EntityRef oldItem) {
+    /**
+     * Changes held item entity.
+     *
+     * <p>Detaches old held item and removes it's components. Adds components to new held item and
+     * attaches it to the mount point entity.</p>
+     */
+    private void linkHeldItemLocationForLocalPlayer(EntityRef newItem, EntityRef oldItem) {
         if (!newItem.equals(oldItem)) {
             EntityRef camera = localPlayer.getCameraEntity();
             FirstPersonHeldItemMountPointComponent mountPointComponent = camera.getComponent(FirstPersonHeldItemMountPointComponent.class);
             if (mountPointComponent != null) {
+
+                if (clientHeldItem.exists()) {
+                    clientHeldItem.destroy();
+                    clientHeldItem = EntityRef.NULL;
+                }
+
+
                 // remove the location from the old item
                 if (oldItem != null && oldItem.exists()) {
-                    Location.removeChild(mountPointComponent.mountPointEntity, oldItem);
-                    oldItem.removeComponent(LocationComponent.class);
                     oldItem.removeComponent(ItemIsHeldComponent.class);
                 } else {
-                    Location.removeChild(mountPointComponent.mountPointEntity, getHandEntity());
-                    getHandEntity().removeComponent(LocationComponent.class);
                     getHandEntity().removeComponent(ItemIsHeldComponent.class);
                 }
 
                 // use the hand if there is no new item
-                EntityRef heldItem = newItem;
-                if (!heldItem.exists()) {
+                EntityRef heldItem;
+                if (!newItem.exists()) {
                     heldItem = getHandEntity();
+                } else {
+                    heldItem = newItem;
```

Comment:
I think so, it references the currently held item which is copied in line 179 to be used as a client-side model. Also in line 188 the `ItemIsHeldComponent` is added to mark the inventory item as being equipped.

The basic idea is to not change any of the previous inventory logic, but create a new client-side entity for the equipped items. So there are two entities, the "inventory entity" (which was replicated to the server and thus shown at 0, 0, 0)  and the new "client-side entity" which is only shown to the current client.
So this code determines the currently held item (hands or inventory item) which is afterwards cloned to the client. This may not be necessary if the hands are equipped, but this way it is consistent and simpler.